File: frontend/lib/models/hoi_dong_phong_van.dart
import 'nguoi_dung.dart';

class ThanhVien {
  final String nguoiDungId;
  final NguoiDung? nguoiDung;
  final String vaiTro;
  final DateTime ngayThamGia;

  ThanhVien({
    required this.nguoiDungId,
    this.nguoiDung,
    required this.vaiTro,
    required this.ngayThamGia,
  });

  factory ThanhVien.fromJson(Map<String, dynamic> json) {
    return ThanhVien(
      nguoiDungId: json['nguoiDungId'] is String
          ? json['nguoiDungId']
          : json['nguoiDungId']?['_id'] ?? '',
      nguoiDung: json['nguoiDungId'] is Map
          ? NguoiDung.fromJson(json['nguoiDungId'])
          : null,
      vaiTro: json['vaiTro'] ?? 'Thanh_vien',
      ngayThamGia: DateTime.parse(
        json['ngayThamGia'] ?? DateTime.now().toIso8601String(),
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'nguoiDungId': nguoiDungId,
      'vaiTro': vaiTro,
      'ngayThamGia': ngayThamGia.toIso8601String(),
    };
  }

  String get vaiTroText {
    const map = {
      'Chu_tich': 'Chủ tịch',
      'Thu_ky': 'Thư ký',
      'Thanh_vien': 'Thành viên',
    };
    return map[vaiTro] ?? vaiTro;
  }
}

class HoiDongPhongVan {
  final String id;
  final String ten;
  final String? moTa;
  final List<ThanhVien> thanhVien;
  final String trangThai;
  final String taoBoi;
  final int soLuocPhongVan;
  final DateTime createdAt;
  final DateTime updatedAt;

  HoiDongPhongVan({
    required this.id,
    required this.ten,
    this.moTa,
    required this.thanhVien,
    required this.trangThai,
    required this.taoBoi,
    this.soLuocPhongVan = 0,
    required this.createdAt,
    required this.updatedAt,
  });

  factory HoiDongPhongVan.fromJson(Map<String, dynamic> json) {
    return HoiDongPhongVan(
      id: json['_id'] ?? '',
      ten: json['ten'] ?? '',
      moTa: json['moTa'],
      thanhVien: (json['thanhVien'] as List? ?? [])
          .map((e) => ThanhVien.fromJson(e))
          .toList(),
      trangThai: json['trangThai'] ?? 'Hoat_dong',
      taoBoi: json['taoBoi'] ?? '',
      soLuocPhongVan: json['soLuocPhongVan'] ?? 0,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'ten': ten,
      'moTa': moTa,
      'thanhVien': thanhVien.map((e) => e.toJson()).toList(),
      'trangThai': trangThai,
      'taoBoi': taoBoi,
      'soLuocPhongVan': soLuocPhongVan,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  int get soThanhVien => thanhVien.length;
  bool get isActive => trangThai == 'Hoat_dong';
}

File: backend/src/models/HoiDongPhongVan.js
const mongoose = require('mongoose');
const { VAI_TRO_HOI_DONG } = require('../utils/constants');

const thanhVienSchema = new mongoose.Schema({
  nguoiDungId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung',
    required: true
  },
  vaiTro: {
    type: String,
    enum: Object.values(VAI_TRO_HOI_DONG),
    required: true
  },
  ngayThamGia: {
    type: Date,
    default: Date.now
  }
}, { _id: false });

const hoiDongPhongVanSchema = new mongoose.Schema({
  ten: {
    type: String,
    required: [true, 'Vui lòng nhập tên hội đồng'],
    trim: true,
    maxlength: [200, 'Tên hội đồng không được quá 200 ký tự']
  },
  moTa: {
    type: String,
    trim: true
  },
  thanhVien: {
    type: [thanhVienSchema],
    required: true,
    validate: {
      validator: function(v) {
        return v && v.length > 0;
      },
      message: 'Hội đồng phải có ít nhất 1 thành viên'
    }
  },
  trangThai: {
    type: String,
    enum: ['Hoat_dong', 'Khong_hoat_dong'],
    default: 'Hoat_dong'
  },
  taoBoi: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung',
    required: true
  },
  soLuocPhongVan: {
    type: Number,
    default: 0
  }
}, {
  timestamps: true,
  collection: 'hoi_dong_phong_van'
});

// Index
hoiDongPhongVanSchema.index({ ten: 1 });
hoiDongPhongVanSchema.index({ trangThai: 1 });
hoiDongPhongVanSchema.index({ 'thanhVien.nguoiDungId': 1 });

// Virtual để đếm số thành viên
hoiDongPhongVanSchema.virtual('soThanhVien').get(function() {
  return this.thanhVien.length;
});

// Validate: Chỉ có 1 chủ tịch
hoiDongPhongVanSchema.pre('save', function(next) {
  const chuTichCount = this.thanhVien.filter(
    tv => tv.vaiTro === VAI_TRO_HOI_DONG.CHU_TICH
  ).length;
  
  if (chuTichCount > 1) {
    return next(new Error('Hội đồng chỉ có thể có 1 chủ tịch'));
  }
  
  // Check trùng thành viên
  const nguoiDungIds = this.thanhVien.map(tv => tv.nguoiDungId.toString());
  const uniqueIds = new Set(nguoiDungIds);
  
  if (nguoiDungIds.length !== uniqueIds.size) {
    return next(new Error('Không thể thêm trùng thành viên trong hội đồng'));
  }
  
  next();
});

// Method thêm thành viên
hoiDongPhongVanSchema.methods.themThanhVien = function(nguoiDungId, vaiTro) {
  const exists = this.thanhVien.some(
    tv => tv.nguoiDungId.toString() === nguoiDungId.toString()
  );
  
  if (exists) {
    throw new Error('Thành viên đã tồn tại trong hội đồng');
  }
  
  this.thanhVien.push({ nguoiDungId, vaiTro });
};

// Method xóa thành viên
hoiDongPhongVanSchema.methods.xoaThanhVien = function(nguoiDungId) {
  this.thanhVien = this.thanhVien.filter(
    tv => tv.nguoiDungId.toString() !== nguoiDungId.toString()
  );
  
  if (this.thanhVien.length === 0) {
    throw new Error('Hội đồng phải có ít nhất 1 thành viên');
  }
};

const HoiDongPhongVan = mongoose.model('HoiDongPhongVan', hoiDongPhongVanSchema);

module.exports = HoiDongPhongVan;